### PR TITLE
Fix urdfreader and luamodel addons for Casadi backend.

### PR DESCRIPTION
**Pull request by PeterQLee**
Here's an update to allow the urdfreader and luamodel addons to be run with casadi. If the user specifies both the `RBDL_USE_CASADI_MATH` and `RBDL_BUILD_ADDON_URDFREADER` / `RBDL_BUILD_ADDON_LUAMODEL` flags, additional libraries (librbdl-casadi_urdfreader.so / librbdl-casadi_luamodel.so) will be built. 

So if the user wants to build a program using these addons utilizing casadi backend they would link against these seperate libraries (librbdl-casadi.so, librbdl-casadi_urdfreader.so, and librbdl-casadi_luamodel.so).
On the other hand if they want to use the eigen backend they would link against (librbdl.so, librbdl_urdfreader, and librbdl_luamodel.so) as usual.

Note that we didn't test against the "muscle" plugin for the luamodel as it was outside of our focus.

### CODE REVIEW

**Comment by ju6ge** [comment]
This is awesome :muscle: … I will take a deeper look over the weekend. Stay tuned ;)

**Comment by ju6ge** [comment]
Hey, so firstly thanks a lot for the work. I like that you also added some test for the urdfmodels to the addon ;)

But after a closer look I have some complaints, but nothing that can not be resolved. Mostly it has to do with the way you integrated the casadi build of the addons. I sent you a pull request to your fork. Please take a look and test if that works for you.

To summarize my complaints. The build of the casadi addon should not be integrated into the CMakeLists for every addon directly, because this would make the build files even more unreadable due to more if-else-ing. Instead mirror the CMakeList.txt in the casadi directory. This is much cleaner and makes it easier to change all required parts such as adding the headers to the install directories, which was missing ;).

Another thing do not set the definition RBDL_USE_CASADI_MATH directly in the cmake. This may interfere with the correct building and break building both versions of rbdl in parallel. This variable gets set by the rbdl_config.h which differs between the casadi and the normal version of rbdl. For the build to work correctly this rbdl_config.h needs to be included correctly. But this should be handled by the FindCMake files, when building against rbdl/rbdl-casadi.

File: addons/urdfreader/urdfreader.cc [hunk 1]
```diff
@@ -245,10 +237,12 @@ void construct_model(Model *rbdl_model, ModelPtr urdf_model,
...
 
     // compute the joint transformation
+    urdf::Vector3 joint_rpy_temp;
     Vector3d joint_rpy;
     Vector3d joint_translation;
-    urdf_joint->PARENT_TRANSFORM.rotation.RPY(joint_rpy[0],
-                                              joint_rpy[1], joint_rpy[2]);
+    urdf_joint->PARENT_TRANSFORM.rotation.RPY(joint_rpy_temp.x,
+                                              joint_rpy_temp.y, joint_rpy_temp.z);
+    joint_rpy.set(joint_rpy_temp.x, joint_rpy_temp.y, joint_rpy_temp.z);
```
Comment (ju6ge) [inline addons/urdfreader/urdfreader.cc:245]: I am not really sure why this temporary variable would be necessary?

File: addons/urdfreader/urdfreader.cc [hunk 2]
```diff
@@ -280,9 +275,10 @@ void construct_model(Model *rbdl_model, ModelPtr urdf_model,
         I_child->origin.position.x,
         I_child->origin.position.y,
         I_child->origin.position.z);
-      I_child->origin.rotation.RPY(link_inertial_rpy[0],
-                                   link_inertial_rpy[1],
-                                   link_inertial_rpy[2]);
+      I_child->origin.rotation.RPY(link_inertial_rpy_temp.x,
+                                   link_inertial_rpy_temp.y,
+                                   link_inertial_rpy_temp.z);
+      link_inertial_rpy.set(link_inertial_rpy_temp.x, link_inertial_rpy_temp.y, link_inertial_rpy_temp.z);
```
Comment (ju6ge) [inline addons/urdfreader/urdfreader.cc:281]: same here? Why the temporary variable?

File: include/rbdl/Joint.h [hunk 3]
```diff
@@ -481,25 +481,31 @@ struct RBDL_DLLAPI Joint {
...
 #else
     } else if (axis_0 == Math::SpatialVector(0., 0., 1., 0., 0., 0.)) {
 #endif
       mJointType = JointTypeRevoluteZ;
 #ifdef RBDL_USE_CASADI_MATH
-    } else if (axis_0[0].is_zero() && axis_0[1].is_zero() && axis_0[2].is_zero()) {
+    } else if (
+	       fabs(casadi::MX::evalf(axis_0[0]).scalar() < std::numeric_limits<double>::epsilon()) &&
+	       fabs(casadi::MX::evalf(axis_0[1]).scalar() < std::numeric_limits<double>::epsilon()) &&
+	       fabs(casadi::MX::evalf(axis_0[2]).scalar() < std::numeric_limits<double>::epsilon())) {
```
Comment (ju6ge) [inline include/rbdl/Joint.h:508]: I also not sure why this change of condition would be necessary? But I am not a casadi expert, so maybe @pariterre  could you have a look at this?

**Review by ju6ge — CHANGES_REQUESTED** [review]

File: include/rbdl/Joint.h (hunk 3 above)
Comment (pariterre) [reply include/rbdl/Joint.h:508]: Because is_zero() checks if it is a pure 0 (as opposed to a numerical zero which is not possible with casadi). Therefore, it cannot be evaluated the same. is_zero is therefore the right way to test this part

Comment (ju6ge) [reply include/rbdl/Joint.h:508]: okay so this change set is wrong then …

Comment (pariterre) [reply include/rbdl/Joint.h:508]: I've never done it that way (evalf)... But since that MX is necessary a 0 or a 1, it may work. I am not sure how it is better that "is_zero()" though

Comment (ju6ge) [reply include/rbdl/Joint.h:508]: I am all for the code being more readable. And is_null seams less confusing to me. Unless there is a specific reason this was changed @PeterQLee?

Comment (pariterre) [reply include/rbdl/Joint.h:508]: you meant `is_zero`?

Comment (ju6ge) [reply include/rbdl/Joint.h:508]: yes is_zero :laughing:

Comment (PeterQLee) [reply include/rbdl/Joint.h:508]: `is_zero` doesn't actually work. I believe that `is_zero` in Casadi is used to check the sparsity of the vector/matrix, and it seems that if you create a vector in a non-sparse way (e.g., `SpatialVector(1., 0., 0., 0., 0., 0.)`) the condition will fail.

Even in the Casadi code they note that "false negatives" are possible. https://github.com/casadi/casadi/blob/master/casadi/core/mx.hpp#L275

The only definitive way I've figured out how to handle this is to a) do all the logic while the values are still floating point and before they've been converted to a symbolic or b) use evalf to find the floating point value.

Comment (ju6ge) [reply include/rbdl/Joint.h:508]: Okay that settle is then :+1:

File: addons/urdfreader/urdfreader.cc (hunk 2 above)
Comment (PeterQLee) [reply addons/urdfreader/urdfreader.cc:281]: The reason the `_temp` variables were created for this and the above snippets is because the `urdf_joint->PARENT_TRANSFORM.rotation.RPY` command requires floating point (double) references as arguments. This will not compile if you try to use `joint_rpy` in the Casadi version because `Vector3d` in that case is a vector of symbolics (MX). I couldn't think of a cleaner way to do this, but if you have ideas let me know.

**Comment by PeterQLee** [comment]
> Hey, so firstly thanks a lot for the work. I like that you also added some test for the urdfmodels to the addon ;)
> 
> But after a closer look I have some complaints, but nothing that can not be resolved. Mostly it has to do with the way you integrated the casadi build of the addons. I sent you a pull request to your fork. Please take a look and test if that works for you.
> 
> To summarize my complaints. The build of the casadi addon should not be integrated into the CMakeLists for every addon directly, because this would make the build files even more unreadable due to more if-else-ing. Instead mirror the CMakeList.txt in the casadi directory. This is much cleaner and makes it easier to change all required parts such as adding the headers to the install directories, which was missing ;).
> 
> Another thing do not set the definition RBDL_USE_CASADI_MATH directly in the cmake. This may interfere with the correct building and break building both versions of rbdl in parallel. This variable gets set by the rbdl_config.h which differs between the casadi and the normal version of rbdl. For the build to work correctly this rbdl_config.h needs to be included correctly. But this should be handled by the FindCMake files, when building against rbdl/rbdl-casadi.

Thanks for the suggestions. I'm all for cleaning up the CMake configurations haha. I've had some issues compiling with the suggested change, can you take a look? https://github.com/PeterQLee/rbdl/pull/1

File: addons/urdfreader/urdfreader.cc (hunk 2 above)
Comment (ju6ge) [reply addons/urdfreader/urdfreader.cc:281]: okay, maybe add a small comment to make this clear ;)

File: addons/urdfreader/urdfreader.cc (hunk 1 above)
Comment (ju6ge) [reply addons/urdfreader/urdfreader.cc:245]: add comment about mismatch of types regarding symbolic and double ;)

**Comment by ju6ge** [comment]
@PeterQLee I sent a reply to your issue ;)

File: addons/urdfreader/urdfreader.cc (hunk 2 above)
Comment (PeterQLee) [reply addons/urdfreader/urdfreader.cc:284]: Added https://github.com/PeterQLee/rbdl/blob/master/addons/urdfreader/urdfreader.cc#L240-L241

**Comment by PeterQLee** [comment]
Ok, I added the Cmake changes from https://github.com/PeterQLee/rbdl/pull/1

Another change I made was the option to build the casadi build. Where the user used to do `cmake -DRBDL_USE_CASADI_MATH` they should now do `cmake -DRBDL_BUILD_CASADI`. This is because this variable is used the rbdl_config.h https://github.com/PeterQLee/rbdl/blob/master/include/rbdl/rbdl_config.h.cmake#L25 . Without the change the user wouldn't be able to separately build the two eigen and casadi versions of the library.

**Comment by ju6ge** [comment]
I will take a look later today, thanks for your work :muscle:

**Review by ju6ge — APPROVED** [review]

**Comment by ju6ge** [comment]
Okay so this seems to do the trick nicely. At least both versions are building nicely and the test suit is working. So I will merge this, though I still want to do a little deeper inspection of the state after installing the code to the system. So I think I will be merging this sometime in the beginning of next week. Maybe @pariterre could also test if his down stream biorbd will still work as expected …

In the long run if you want to use casadi more, the test suite should also be enabled to test the casadi version. All the existing test should be reusable, just that calling the eval function for the casadi evaluation of the results needs to be handled. But if you want to have more confidence in the results of the casadi version this should absolutly be done! Anyway have a great weekend and thanks for your efforts :muscle:

**Comment by PeterQLee** [comment]
My pleasure, thanks for all of your help in getting this together. 

Enabling tests for casadi is a good idea. I vaguely recall that @pariterre mentioned that he had a suite tests for casadi? Maybe those were specific to biorbd. In any case, the standard rbdl tests would be a good thing to convert; we'll have to look into this at a later date.

**Comment by ju6ge** [comment]
> My pleasure, thanks for all of your help in getting this together.
> 
> Enabling tests for casadi is a good idea. I vaguely recall that @pariterre mentioned that he had a suite tests for casadi? Maybe those were specific to biorbd. In any case, the standard rbdl tests would be a good thing to convert; we'll have to look into this at a later date.

If you do, a good approach would be to create wrapper definitons for the different types of checks the test suit performs. Like testing if a matrix, vector or scalar value are equal. The normal version would just call the normal catch2 check function and the casadi version, which gets selected by the same #ifdef RBDL_USE_CASADI_MATH include guard would first call evalf to evaluate the casadi expression before calling the catch2 check. Then only the check function within the tests would need to be updated to this new wrapper function. Maybe I will just do this when I have some time, but that is limited and if I have more time again the first thing on my RBDL Todo list ist to enable a CI pipeline again ;)

**Comment by pariterre** [comment]
> Enabling tests for casadi is a good idea. I vaguely recall that @pariterre mentioned that he had a suite tests for casadi? Maybe those were specific to biorbd.

@PeterQLee 
This is indeed for biorbd. The approach I used was to create a macro (using a define) that eigen calls the function or creates a casadi function then calls it, depending on which build I am testing (based on the defined for each of the MATH backend). The relevant bits is there https://github.com/pyomeca/biorbd/blob/6150fc2ce05cdca901652be4bbc332eab943d15e/include/biorbdConfig.h.in#L107 . Also, it needs to define the vector a specific way (again depending on which build is tested). I used the following macro https://github.com/pyomeca/biorbd/blob/6150fc2ce05cdca901652be4bbc332eab943d15e/include/biorbdConfig.h.in#L88 

Apart from changing the `BIORBD_NAMESPACE::rigidbody::GeneralizedCoordinates` to `Math::VectorNd` in the vector macro, this should be pretty much be possible to use it as is. Here is an example on how to change the gravity that uses ForwardDynamics to test is the change was effective https://github.com/Sina-Mehdiz/Biorbd/blob/6c7346a5eb9d522db32bb0737cad0cad9a20ca4e/test/test_rigidbody.cpp#L37

**Comment by ju6ge** [comment]
@pariterre Can you test if this PR runs into any issues with your pipeline? Then I will happily merge it ;)

**Comment by pariterre** [comment]
The tests are still passing smoothly in biorbd for both eigen and casadi. No worries here!

**Comment by ju6ge** [comment]
Great then I will merge this. I also checked the installation paths and everything looks great :+1: 

@pariterre should I add a new release for conda?

**Comment by pariterre** [comment]
Do not feel the need, that depends if @PeterQLee installs from conda too
